concrete `cartesian_product(::Zonotope, ::Zonotope)`

Something like

```julia
function LazySets.cartesian_product(Z1::Zonotope, Z2::Zonotope)
    center = vcat(Z1.center, Z2.center)
    z1 = zeros(dim(Z1), size(Z2.generators, 2))
    z2 = zeros(dim(Z2), size(Z1.generators, 2))
    g1 = Z1.generators
    g2 = Z2.generators
    generators = [
        g1 z1;
        g2 z2
    ]
    return Zonotope(center, generators)
end
```

I wasn't sure if this was an omission, or intentionally unimplemented

Seems like a missing method. The idea you proposed lgtm.

There is a fallback, but it is suboptimal since it passes to the hrep:

```julia
julia> using LazySets

julia> Z = rand(Zonotope, dim=2);

julia> Y = rand(Zonotope, dim=2);

julia> cartesian_product(Z, Y)
ERROR: AssertionError: package 'Polyhedra' not loaded (it is required for executing ``cartesian_product`)
.....

julia> using Polyhedra

julia> cartesian_product(Z, Y)
HPolyhedron{Float64, Vector{Float64}}(LazySets.HalfSpace{Float64, Vector{Float64}}[LazySets.HalfSpace .....
```

--- a/src/ConcreteOperations/cartesian_product.jl
+++ b/src/ConcreteOperations/cartesian_product.jl
@@ -151,3 +151,20 @@ function cartesian_product(S1::AbstractSingleton, S2::AbstractSingleton)
     v2 = element(S2)
     return Singleton(vcat(v1, v2))
 end
+
+function cartesian_product(Z1::AbstractZonotope, Z2::AbstractZonotope)
+    N = promote_type(eltype(Z1), eltype(Z2))
+    z1 = zeros(N, dim(Z1), ngens(Z2))
+    z2 = zeros(N, dim(Z2), ngens(Z1))
+    g1 = genmat(Z1)
+    g2 = genmat(Z2)
+    c = vcat(center(Z1), center(Z2))
+    G = [g1 z1; z2 g2]
+    return Zonotope(c, G)
+end
+
+function cartesian_product(H1::AbstractHyperrectangle, H2::AbstractHyperrectangle)
+    c = vcat(center(H1), center(H2))
+    r = vcat(radius_hyperrectangle(H1), radius_hyperrectangle(H2))
+    return Hyperrectangle(c, r)
+end

--- a/test/ConcreteOperations/cartesian_product.jl
+++ b/test/ConcreteOperations/cartesian_product.jl
@@ -1,0 +1,28 @@
+for N in [Float64, Float32, Rational{Int}]
+    c1 = N[1, 2]
+    c2 = N[3, 4, 5]
+    Z1 = Zonotope(c1, N[1 0; 0 1])
+    Z2 = Zonotope(c2, N[1 0 0; 0 1 0; 0 0 1])
+    H1 = Hyperrectangle(c1, N[1, 1])
+    H2 = Hyperrectangle(c2, N[1, 1, 1])
+    B1 = BallInf(c1, N(1))
+    B2 = BallInf(c2, N(1))
+
+    Z1Z2 = cartesian_product(Z1, Z2)
+    Z1H2 = cartesian_product(Z1, H2)
+    H1Z2 = cartesian_product(H1, Z2)
+    H1H2 = cartesian_product(H1, H2)
+    H1B2 = cartesian_product(H1, B2)
+    B1B2 = cartesian_product(B1, B2)
+    for X in [Z1Z2, Z1H2, H1Z2]
+        @test X isa Zonotope{N}
+    end
+    for X in [H1H2, H1B2, B1B2]
+        @test X isa Hyperrectangle{N}
+    end
+    if test_suite_polyhedra || N <: AbstractFloat
+        for X in [Z1Z2, Z1H2, H1Z2, H1H2, H1B2]
+            @test isequivalent(X, B1B2)
+        end
+    end
+end

--- a/test/runtests.jl
+++ b/test/runtests.jl
@@ -145,6 +145,7 @@ if test_suite_basic
     # Concrete operations
     # ===================
     @time @testset "LazySets.concrete_convex_hull" begin include("ConcreteOperations/convex_hull.jl") end
+    @time @testset "LazySets.concrete_cartesian_product" begin include("ConcreteOperations/cartesian_product.jl") end
     @time @testset "LazySets.samples" begin include("ConcreteOperations/samples.jl") end
     @time @testset "LazySets.area" begin include("ConcreteOperations/area.jl") end
 

